src/nupic/encoders/ScalarEncoder.hpp

```diff
@@ -21,167 +23,119 @@
-
-  protected:
-    const int w_;
-    int n_;
+    double  radius = 0.0f;
+
+    /**
+     * Member "resolution" Two inputs separated by greater than, or equal to the
+     * resolution are guaranteed to have different representations.
+     */
+    double  resolution = 0.0f;
   };
 
-/** Encodes a floating point number as a contiguous block of 1s.
- *
- * @b Description
- * A ScalarEncoder encodes a numeric (floating point) value into an array
- * of bits. The output is 0's except for a contiguous block of 1's. The
- * location of this contiguous block varies continuously with the input value.
- *
- * Conceptually, the set of possible outputs is a set of "buckets". If there
- * are m buckets, the ScalarEncoder distributes m points along the domain
- * [minValue, maxValue], including the endpoints. To figure out the bucket
- * index of an input, it rounds the input to the nearest of these points.
- *
- * This approach is different from the PeriodicScalarEncoder because two
- * buckets, the first and last, are half as wide as the rest, since fewer
- * numbers in the input domain will round to these endpoints. This behavior
- * makes sense because, for example, with the input space [1, 10] and 10
- * buckets, 1.49 is in the first bucket and 1.51 is in the second.
- */
-class ScalarEncoder : public ScalarEncoderBase {
-public:
   /**
-   * Constructs a ScalarEncoder
-   *
-   * @param w The number of bits that are set to encode a single value -- the
-   *   "width" of the output signal
-   * @param minValue The minimum value of the input signal, inclusive.
-   * @param maxValue The maximum value of the input signal, inclusive.
-   * @param clipInput Whether to allow input values outside the [minValue,
-   * maxValue] range.  If true, the input will be clipped to minValue or
-   * maxValue.
+   * Encodes a real number as a contiguous block of 1s.
    *
-   * There are three mutually exclusive parameters that determine the overall
-   * size of of the output. Only one of these should be nonzero:
-   *
-   * @param n The number of bits in the output. Must be greater than or equal to
-   * w.
-   * @param radius Two inputs separated by more than the radius have
-   *   non-overlapping representations. Two inputs separated by less than the
-   *   radius will in general overlap in at least some of their bits. You can
-   *   think of this as the radius of the input.
-   * @param resolution Two inputs separated by greater than, or equal to the
-   *   resolution are guaranteed to have different representations.
+   * Description:
+   * The ScalarEncoder encodes a numeric (floating point) value into an array
+   * of bits. The output is 0's except for a contiguous block of 1's. The
+   * location of this contiguous block varies continuously with the input value.
    */
-  ScalarEncoder(int w, double minValue, double maxValue, int n, double radius,
-                double resolution, bool clipInput);
+  class ScalarEncoder : public BaseEncoder<double>
```

Comment:
use a variant of `Real`

Reply:
and consider allowing also ScalarEncoder to be templated, 
`auto intEncoder = ScalarEncoder<Int>(..);`
and we could provide the default specification `typedef ScalarEncoder<Real> ScalarEncoder;`

Reply:
Ok, will use `Real64`.

I don't want to use templates. The reason to use templates would be to make it more general purpose.  64-bit floating point numbers are also very general purpose because they can represent any number the user might give us. Real64 is simpler to use than templates.  

Performance wise, there should be no measurable difference because the input is used in a half-dozen simple equations.

Reply:
> The reason to use templates would be to make it more general purpose 

And also to make the code more specific where requested, as in my example

> auto intEncoder = ScalarEncoder<Int>(..); 

I want to limit said encoder to integers only, if I then try to encode 3.141, I'd get an error. 

On the other hand, should the code be much simpler with the hard-coded Real, that should be enough.

Reply:
EDIT: GH hides <  and "the other", and text inside, which is really confusing in these comments

Reply:
> I want to limit said encoder to integers only, if I then try to encode 3.141, I'd get an error.

Actually I think in this situation 3.141 would silently truncate to 3.

Reply:
> would silently truncate to 3.

We'd get a warning -> error. But true, let's keep it real :)